File: courts/suites/data-abi/allocator-default-probe.c
/*
 * ALLOCATOR-DEFAULT-001 — differential probe of the DEFAULT allocator
 * contract (11.1-Z.3, R-000178).
 *
 * Compiled twice: against the oracle DSO (system libxml2, whose default
 * `xmlMalloc`/`xmlFree`/`xmlRealloc`/`xmlMemStrdup` variables are the plain
 * C runtime functions — globals.c) and against the candidate DSO
 * (liblibxml_rs, whose `*Default` bodies are libc wrappers). The output is
 * deterministic: it prints only RELATIVE facts (NULL/non-NULL patterns,
 * content preservation, routing counters, and the exact values of
 * xmlMemSize/xmlMemUsed/xmlMemBlocks, which the oracle keeps at 0 while the
 * default allocator is installed) — never pointer values or addresses.
 *
 * The two program outputs must be byte-identical.
 *
 * Exercises:
 *   (1) many allocation sizes + zero-size allocation (C malloc contract);
 *   (2) grow realloc / shrink realloc / realloc-to-zero (C realloc
 *       contract: realloc(p,0) frees and returns NULL on glibc);
 *   (3) realloc(NULL, n) == malloc; realloc failure leaves the old block
 *       intact; malloc(SIZE_MAX) == NULL;
 *   (4) strdup content + strdup(NULL) == NULL; free(NULL) no-op;
 *   (5) long allocation/free churn (100k blocks);
 *   (6) direct exported-variable assignment (xmlMalloc = custom) routes
 *       actual allocations, and the debug counters stay 0 (upstream: custom
 *       hooks never touch debugMemSize);
 *   (7) xmlMemSize/xmlMemUsed/xmlMemBlocks exactness under the default
 *       allocator (all 0, matching the oracle's plain-malloc default);
 *   (8) the display entry points (xmlMemDisplay/xmlMemDisplayLast/xmlMemShow)
 *       write nothing (upstream 2.15.0 removed the feature).
 */
#define _POSIX_C_SOURCE 200809L
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <libxml/xmlmemory.h>

static int nmalloc = 0, nfree = 0;
static void *counting_malloc(size_t s) { nmalloc++; return malloc(s); }
static void counting_free(void *p) { nfree++; free(p); }

/* Suppress -Walloc-size-larger-than for the intentional SIZE_MAX probes. */
static size_t huge_size(void) { return (size_t)-1; }

int main(void) {
    void *p, *q;
    unsigned char buf[64];
    unsigned long i;
    FILE *f;
    long pos;

    /* (1) counters start at 0 with the default allocator (debugMemSize). */
    printf("stats0 used=%d blocks=%d\n", xmlMemUsed(), xmlMemBlocks());

    /* (2) many allocation sizes + zero-size. */
    for (i = 1; i <= 16384; i <<= 1) {
        p = xmlMalloc((size_t)i);
        printf("m%lu nonnull=%d\n", i, p != NULL);
        if (p) xmlFree(p);
    }
    p = xmlMalloc(0);
    printf("m0 nonnull=%d\n", p != NULL);
    if (p) xmlFree(p);

    /* (3) grow realloc preserves content. */
    memset(buf, 0x5a, sizeof(buf));
    p = xmlMalloc(64);
    if (p) memcpy(p, buf, 64);
    q = xmlRealloc(p, 256);
    printf("realloc-grow nonnull=%d content=%d\n",
           q != NULL, q != NULL && memcmp(q, buf, 64) == 0);

    /* (4) shrink realloc preserves the leading content. */
    p = xmlRealloc(q, 16);
    printf("realloc-shrink nonnull=%d content=%d\n",
           p != NULL, p != NULL && memcmp(p, buf, 16) == 0);

    /* (5) realloc(p, 0): glibc frees the block and returns NULL. */
    q = xmlRealloc(p, 0);
    printf("realloc-zero null=%d\n", q == NULL);

    /* (6) realloc(NULL, n) == malloc. */
    p = xmlRealloc(NULL, 32);
    printf("realloc-null nonnull=%d\n", p != NULL);
    xmlFree(p);

    /* (7) realloc failure: old block stays allocated and intact. */
    memset(buf, 0x3c, sizeof(buf));
    p = xmlMalloc(16);
    if (p) memcpy(p, buf, 16);
    q = xmlRealloc(p, huge_size());
    printf("realloc-fail null=%d old-intact=%d content=%d\n",
           q == NULL, q == NULL, p != NULL && memcmp(p, buf, 16) == 0);
    xmlFree(p);

    /* (8) malloc failure. */
    p = xmlMalloc(huge_size());
    printf("malloc-huge null=%d\n", p == NULL);

    /* (9) strdup content + NULL handling (upstream xmlPosixStrdup). */
    p = xmlMemStrdup("hello allocator");
    printf("strdup ok=%d\n", p != NULL && strcmp((char *)p, "hello allocator") == 0);
    xmlFree(p);
    p = xmlMemStrdup(NULL);
    printf("strdup-null null=%d\n", p == NULL);

    /* (10) free(NULL) is a no-op. */
    xmlFree(NULL);
    printf("free-null ok=1\n");

    /* (11) long allocation/free churn (100k blocks, mixed sizes). */
    for (i = 0; i < 100000; i++) {
        p = xmlMalloc(1 + (i % 128));
        if (!p) break;
        *(volatile unsigned char *)p = (unsigned char)i;
        xmlFree(p);
    }
    printf("churn done=%d\n", i == 100000);

    /* (12) direct exported-variable assignment routes actual allocations. */
    nmalloc = nfree = 0;
    xmlMalloc = counting_malloc;
    xmlFree = counting_free;
    p = xmlMalloc(48);
    printf("direct routed=%d nonnull=%d\n", nmalloc == 1, p != NULL);
    xmlFree(p);
    printf("direct free=%d\n", nfree == 1);

    /* (13) exactness: with the default allocator the debug counters stay 0
     * even after hundreds of allocations, and xmlMemSize is 0 for
     * default-allocator blocks (upstream: plain malloc carries no MEMHDR). */
    printf("stats-final used=%d blocks=%d\n", xmlMemUsed(), xmlMemBlocks());
    p = xmlMalloc(80);
    printf("size-live=%zu size-null=%zu\n", xmlMemSize(p), xmlMemSize(NULL));
    xmlFree(p);

    /* (14) display entry points write nothing (upstream 2.15.0 no-ops). */
    f = tmpfile();
    xmlMemDisplay(f);
    xmlMemDisplayLast(f, -1);
    xmlMemShow(f, 10);
    pos = f ? ftell(f) : -1;
    printf("display-bytes=%ld\n", pos);
    if (f) fclose(f);

    return 0;
}

File: tools/abi/allocator_default_probe.py
#!/usr/bin/env python3
"""ALLOCATOR-DEFAULT-001 — differential court for the DEFAULT allocator
contract (11.1-Z.3, R-000178).

Compiles courts/suites/data-abi/allocator-default-probe.c twice:

  1. oracle  — against the system libxml2 (/usr/include/libxml2, -lxml2)
  2. candidate — against the candidate headers (include/) and DSO
               (target/debug/liblibxml_rs.so)

and requires the two program outputs to be byte-identical. The probe prints
only relative facts (NULL patterns, content preservation, routing counters,
and the exact xmlMemSize/xmlMemUsed/xmlMemBlocks values, which must be 0
under the default allocator on both sides), so the two runs must match
exactly.

This court proves the R-000178 default-allocator contract: the candidate's
default hooks are plain libc malloc/realloc/free/strdup wrappers (the
pre-Z.3 implementation used Rust `std::alloc` with fabricated layouts —
invalid-layout UB) with observable behavior identical to the oracle's
globals.c defaults (`xmlMalloc = malloc` etc.):
  - malloc(0) non-NULL, malloc(SIZE_MAX) NULL;
  - realloc grow/shrink preserve content, realloc(p,0) frees+returns NULL,
    realloc(NULL,n) allocates, realloc failure leaves the old block;
  - strdup content + strdup(NULL) NULL; free(NULL) no-op;
  - 100k alloc/free churn;
  - direct exported-variable assignment routes allocations;
  - xmlMemUsed/xmlMemBlocks/xmlMemSize stay 0 (upstream debug counters are
    maintained only by the debug allocator).

When valgrind is available, the candidate probe binary is additionally run
under valgrind (--error-exitcode) as a memory-safety sweep; the result is
recorded in the receipt but the verdict is the differential equality.

Receipts are written to courts/receipts/phase-11/.

Usage:
    python3 tools/abi/allocator_default_probe.py
"""

import datetime
import hashlib
import json
import os
import shutil
import subprocess
import sys

ROOT = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
PROBE = os.path.join(ROOT, "courts", "suites", "data-abi", "allocator-default-probe.c")
RECEIPTS = os.path.join(ROOT, "courts", "receipts", "phase-11")
CAND_SO = os.path.join(ROOT, "target", "debug", "liblibxml_rs.so")


def run(cmd, timeout=120):
    r = subprocess.run(cmd, capture_output=True, text=True, timeout=timeout)
    return r.returncode, r.stdout, r.stderr


def main():
    os.makedirs(RECEIPTS, exist_ok=True)
    for f in (PROBE, CAND_SO):
        if not os.path.exists(f):
            print(f"MISSING {f}")
            return 1

    # ---- oracle build (system headers) ------------------------------------
    rc, _, err = run(["gcc", "-std=c11", "-o", os.path.join(ROOT, "target", "ad-oracle"),
                      PROBE, "-I", "/usr/include/libxml2", "-lxml2"])
    if rc != 0:
        print("ORACLE COMPILE FAILED:\n" + err[-2000:])
        return 1
    rc, o_out, err = run([os.path.join(ROOT, "target", "ad-oracle")])
    if rc != 0:
        print("ORACLE RUN FAILED:\n" + err[-2000:])
        return 1

    # ---- candidate build (candidate headers + DSO) ------------------------
    rc, _, err = run(["gcc", "-std=c11", "-o", os.path.join(ROOT, "target", "ad-cand"),
                      PROBE, "-I", os.path.join(ROOT, "include"),
                      "-L", os.path.join(ROOT, "target", "debug"), "-llibxml_rs",
                      "-Wl,-rpath," + os.path.join(ROOT, "target", "debug")])
    if rc != 0:
        print("CANDIDATE COMPILE FAILED:\n" + err[-2000:])
        return 1
    rc, c_out, err = run([os.path.join(ROOT, "target", "ad-cand")])
    if rc != 0:
        print("CANDIDATE RUN FAILED:\n" + err[-2000:])
        return 1

    oracle_hash = hashlib.sha256(o_out.encode()).hexdigest()
    cand_hash = hashlib.sha256(c_out.encode()).hexdigest()
    identical = oracle_hash == cand_hash

    mismatch_lines = []
    if not identical:
        o_lines, c_lines = o_out.splitlines(), c_out.splitlines()
        for i, (a, b) in enumerate(zip(o_lines, c_lines)):
            if a != b:
                mismatch_lines.append((i + 1, a, b))
        if len(o_lines) != len(c_lines):
            mismatch_lines.append(("len", len(o_lines), len(c_lines)))

    # ---- optional valgrind sweep on the candidate (memory safety) ---------
    # NB: valgrind 3.25.1 in this environment SIGILLs inside the dynamic
    # loader (_dl_start) on every binary, including /bin/true — an
    # environment/toolchain incompatibility, not a probe failure. The court
    # records `usable: false` with the reason when that is observed.
    valgrind = shutil.which("valgrind")
    vg = None
    if valgrind:
        vg_rc, _, vg_err = run(
            [valgrind, "--error-exitcode=99", "--leak-check=full",
             "--errors-for-leak-kinds=definite,indirect", "-q",
             os.path.join(ROOT, "target", "ad-cand")],
            timeout=600,
        )
        loader_crash = "_dl_start" in vg_err and "SIGILL" in vg_err
        vg = {
            "available": True,
            "binary": valgrind,
            "error_exitcode": 99,
            "exit_code": vg_rc,
            "usable": not loader_crash and vg_rc == 0,
            "reason": "valgrind SIGILLs in the dynamic loader (_dl_start) on every "
            "binary in this environment (toolchain incompatibility); the "
            "memory-safety sweep is therefore not executable here"
            if loader_crash
            else None,
            "stderr_tail": vg_err[-2000:],
        }
    else:
        vg = {"available": False}

    ts = datetime.datetime.now(datetime.timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    receipt = {
        "court": "ALLOCATOR-DEFAULT",
        "case": "ALLOCATOR-DEFAULT-001",
        "phase": "11.1-Z.3",
        "timestamp": ts,
        "schema": "allocator-default-differential-1",
        "probe": os.path.relpath(PROBE, ROOT),
        "oracle_sha256": oracle_hash,
        "candidate_sha256": cand_hash,
        "identical": identical,
        "mismatches": mismatch_lines[:40],
        "valgrind": vg,
        "verdict": "PASS" if identical else "FAIL",
    }
    rp = os.path.join(RECEIPTS, f"allocator-default-{ts}.json")
    with open(rp, "w") as f:
        json.dump(receipt, f, indent=1, ensure_ascii=False)
        f.write("\n")
    print(f"receipt -> {rp}")
    print(f"oracle={oracle_hash} candidate={cand_hash} identical={identical}")
    if valgrind:
        print(f"valgrind exit={vg['exit_code']} usable={vg['usable']}")
    if not identical:
        print("--- first mismatches:")
        for ln, a, b in mismatch_lines[:20]:
            print(f"  L{ln}: oracle  {a!r}\n       candidate {b!r}")
    return 0 if identical else 1


if __name__ == "__main__":
    sys.exit(main())
